### ISSUE TITLE
#include <stdint.h> in gmp.h/mpir.h on Solaris?

### ISSUE THREAD

**Issue by dimpase** [wbhart/mpir#242]
While building Sage on SPARC Solaris (hey, no Spectre ;-))
it appears that `#include <stdint.h>` is needed in `gmp.h`, otherwise `./configure` of MPFR is failing with `unknown type name 'uintmax_t'` and  `'intmax_t'`.
And this is fixed by adding `#include <stdint.h>` somewhere near the top of  MPIR's `gmp.h` (the latter is a copy of `mpir.h`, needless to say).

Well, it might be it's an MPFR bug, I'm just trying to understand what's exactly happening here, and why MPIR does not include this header.

**Comment by wbhart** [comment]
I doubt it's just a simple matter of including it or not. Isn't there a
configure test?

**Comment by dimpase** [comment]
Yes, there is a configure test, specifically, checking for GMP_NUMB_BITS
and sizeof(mp_limb_t) consistency

and it, and the whole configure process, fails, with the message
that these are not consistent.

However looking at the config.log tells you that it actually fails to
compile this test, with errors as I reported; and it goes through with the
hack I reported.

**Comment by dimpase** [comment]
That is, it's entirely possible that MPFR misses a test for intmax_t (in AC_TYPE_INTMAX_T?), etc,
I didn't look at it yet, I am just not sure at which direction to dig...

**Comment by wbhart** [comment]
Unless it's something SPARC specific, it's unlikely to be an issue with
MPIR. Anyway, I don't think I can help you. I don't have a clue what might
be going on.

**Comment by dimpase** [comment]
You somehow assume that if INTMAX_MAX is defined then you can use uintmax_t and intmax_t.
Why? Note that the error comes from lines in gmp.h.in : 
```
/* if stdint.h is available -- n.b: we do NOT include stdint.h ourselves */
#if defined(INTMAX_MAX)

#define __GMP_BITS_PER_UINTMAX  (8*sizeof(uintmax_t))
...
```
I could not find anything meaningful about INTMAX_MAX. IMHO you rely on some unwritten convention that works for some reason in most cases, something like this. Why not instead check for the presence of this header and include it?

**Comment by wbhart** [comment]
We assume it because if it is available then uintmax_t is defined on the
system by stdint.h (which it is assumed has been included at this point),
according to the standard.

http://pubs.opengroup.org/onlinepubs/009695399/basedefs/stdint.h.html

I don't see how this is relevent to deciding if stdint.h is available on
the system. This comes after that has already been decided.

**Comment by dimpase** [comment]
The link (somehow Google didn't find it for me...) merely says that INTMAX_MAX specifies
the maximum of a type  declared in the stdint.h header. What does make you think that it somehow ensures that this header is included at the point where INTMAX_MAX is available?
And the following is a quote from mpfr.h
```
We do not check INTMAX_MAX and UINTMAX_MAX because under Solaris,
       these macros are always defined by <limits.h> (i.e. even when
       <stdint.h> and <inttypes.h> are not included).
```

By the way, a newer link is http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/stdint.h.html
but it does not seem to be changed (much).

One way or another, this is not a standard. It's said right at the beginning that it is a standard extension.

**Comment by wbhart** [comment]
So Solaris defines INTMAX_MAX without defining intmax_t? I have no sympathy
for them.

**Comment by jdemeyer** [comment]
> So Solaris defines INTMAX_MAX without defining intmax_t?

Yes

**Comment by wbhart** [comment]
Patches are welcome. But just unconditionally including stdint.h is not the
way to fix this. The logic that is there is there for a reason. We need to
know on which machines intmax_t is available and which it isn't since
compilation fails on machines it is not available on.

**Comment by wbhart** [comment]
Please also see the following from the MPIR 2.7.0 release announcement:

* Important note: mpz_{set,get}_{ux,sx} are now only available if stdint.h is
included before gmp/mpir.h or gmp/mpirxx.h.

So this was done deliberately as the only solution to a difficult issue.
There was considerable discussion between the MPIR devs about this.

**Comment by dimpase** [comment]
Yes, certainly one should check that the header exists, and include it if it does.
The logic of  `we do NOT include stdint.h ourselves ` is beyond me.

I think if you use a type, make sure it is defined, do not leave it to others as a cocked hammer to shoot themselves in the foot.

**Comment by wbhart** [comment]
That's not what's going on here. There was a long, involved discussion
about this, as we were fixing a difficult issue with MPIR.

The bug here is that INTMAX_MAX is defined when intmax_t is not. That is
indeed awkward, and we should patch for that by testing some other way if
intmax_t is available. But note: it is only available *if* stdint.h has
been included before gmp.h, etc.

This occurs in multiple places in MPIR in different contexts (MPIR
distinguishes between different kinds of builds, it's not a simple linear
list of includes). Different things happen depending on whether gmp.h is
being used during configure, during an MPIR build or by a user of MPIR, or
with/without C++ support. This has always been the case with GMP and MPIR
and is a subtlety that most other libraries do not have to contend with.

**Comment by wbhart** [comment]
Here is an excerpt of what was written to the MPIR list on this topic. Note
that the original patch that was applied at the time didn't actually work.
The Sage people reported that, we fixed it, and that's how we got to where
we are today.

"I am going to change how we deal with the functions mpz_get_ux,
mpz_set_ux, mpz_get_sx, mpz_set_sx.

As you are aware, these require intmax_t to be available, which in
practical terms means that stdint.h needs to be available [sic, apparently].

Determining when stdint.h is available is for the most part easy, due to
the autoconf macro HAVE_STDINT_H, and with MSVC one just looks at the MSVC
version number.

However, the autoconf solution doesn't work for gmp/mpir.h since the latter
does not include config.h (nor can it).

Currently we are checking if _STDINT_H, STDINT_H or _STDINT_H_ are defined.
But this is non-portable. For example, it doesn't work on OpenBSD. So this
has to be changed.

One potential solution would be to have configure substitute macros in gmp-h
.in when creating mpir/gmp.h. But this doesn't work when cross compiling."

In the end, we decided to require stdint.h to be included before gmp.h if
the mpz_set/get_ux/sx functions are desired. You can see there's a very
specific logic in the fix that we applied for this issue. There is no
portable way to determine if intmax_t is available in each possible
scenario. Therefore we *must* require the user to include stdint.h, not do
it automatically ourselves, as this affects gmp.h, which could end up
anywhere, not just the machine it was built on!

**Comment by dimpase** [comment]
Note that `configure.ac` says: https://github.com/wbhart/mpir/blob/429801c6005e43433c9e4d28230a557ee690e97b/configure.ac#L1982

This does not seem to reflect the reality. 

Also, I don't understand why `config.h` is not included in `gmp/mpir.h(.in)`. After all `config.h` is generated by ./configure, so it's irrelevant what happens on the machine where autoconf is run, the real configuration is happening on the machine MPIR is installed.

**Comment by wbhart** [comment]
On 31 January 2018 at 00:20, Dima Pasechnik <notifications@github.com>
wrote:

> Note that configure.ac says: https://github.com/wbhart/mpir/blob/
> 429801c6005e43433c9e4d28230a557ee690e97b/configure.ac#L1982
>
> This does not seem to reflect the reality.
>

I don't see why you think that.


> Also, I don't understand why config.h is not included in gmp/mpir.h(.in).
> After all config.h is generated by ./configure, so it's irrelevant what
> happens on the machine where autoconf is run, the real configuration is
> happening on the machine MPIR is installed.
>

MPIR cross-compiles.

**Comment by dimpase** [comment]
For cross-compilation, you need headers of the the system you build for; `./configure` will analyse headers for the "host" system, and generate correct for the task `config.h`.

**Comment by wbhart** [comment]
On 31 January 2018 at 01:10, Dima Pasechnik <notifications@github.com>
wrote:

> For cross-compilation, you need headers of the the system you build for;
> ./configure will analyse headers for the "host" system, and generate
> correct for the task config.h.
>

You don't necessarily know what will be on the host system. In some cases,
you have access to the tools that are available on that system, and
configure supports this. In other cases, you are building for a class of
systems with a different architecture. Then you don't have access to
anything. You certainly don't have access to the headers, even if that were
a consistent set of files.

gmp.h has to be valid for *any* system. It can't rely on stuff supplied by
autoconf. This is by design.

What one would need to solve the problem the way you want to solve it, is a
way for gmp.h to detect if intmax_t (or stdint.h) is available, at compile
time, regardless of where it is being #included. Since that's not actually
possible, you have to assume stdint.h was included before gmp.h if the user
wants to use stuff from it. There's really no way around this.

**Comment by wbhart** [comment]
Well, I correct myself a little: gmp.h has to be valid for *any* system
with the specified ABI, microarchitecture, OS and vendor.

**Comment by wbhart** [comment]
Also consider fat binaries. The deployment of MPIR can be very wide for a
given build.

**Comment by dimpase** [comment]
I don't understand what kind of setup you keep in mind - if you cross-compile for a system, it's irrelevant what headers that actual system has, as it has no tools to build anything anyway (otherwise you're better off using tools there...)

Such issues only arise in closed-source scenarios, not in open-source ones.

**Comment by wbhart** [comment]
On 31 January 2018 at 01:53, Dima Pasechnik <notifications@github.com>
wrote:

> I don't understand what kind of setup you keep in mind - if you
> cross-compile for a system, it's irrelevant what headers that actual system
> has, as it has no tools to build anything anyway (otherwise you're better
> off using tools there...)
>
> Such issues only arise in closed-source scenarios, not in open-source ones.
>
GMP itself cross-compiles, and you couldn't find a project more committed
to Open Source. And it's a library, so what do you think people are doing
with it?

The changes that were made were because of an actual real world scenario.
We didn't make the changes just for the hell of it.

I don't think you are seeing the wide range of places and ways this
software gets deployed.

**Comment by wbhart** [comment]
For an example of requiring headers to be included before gmp.h to enable
some functionality in GMP, see section 3.1 of the GMP manual [1].

[1] https://gmplib.org/gmp-man-6.1.2.pdf

**Comment by dimpase** [comment]
It could be just a historical artefact from pre-autoconf times of GMP, nothing more than that.
(I wrote my 1st code using GMP in 1994, mind you :-))

**Comment by dimpase** [comment]
fixed by #245 
Closing this?
